### PR TITLE
Add all_flags, secure_mode_hash + more

### PR DESCRIPTION
**Pull request by drichelson**
Getting client up to speed with: https://catamorphic.quip.com/Ggk6AbylpJBf

### CODE REVIEW

File: ldclient/flag.py [hunk 1]
```diff
@@ -14,29 +14,46 @@
 log = logging.getLogger(sys.modules[__name__].__name__)
 
 
-def evaluate(flag, user, store, prereq_events=[]):
+def evaluate(flag, user, store):
```
Comment (drichelson) [inline ldclient/flag.py]: I moved all the flag eval logic into the flag file.

File: ldclient/flag.py [hunk 2]
```diff
@@ -14,29 +14,46 @@
...
+        if value is not None:
+            return value, prereq_events
+
+    if 'offVariation' in flag and flag['offVariation']:
+        value = _get_variation(flag, flag['offVariation'])
+        return value, prereq_events
+    return None, prereq_events
+
+
+def _evaluate(flag, user, store, prereq_events=None):
```
Comment (drichelson) [inline ldclient/flag.py]: When we had the default value for prereq_events set to [] it was getting reused across function calls and the array just kept getting appended and we sent back a ton of prereq events.

File: demo/demo.py [hunk 3]
```diff
@@ -15,11 +15,11 @@
 root.addHandler(ch)
 
 if __name__ == '__main__':
-    ldclient._api_key = 'api_key'
+    ldclient.api_key = 'api_key'
```
Comment (jkodumal) [inline demo/demo.py]: Can we change api_key to `sdk_key`?

**Comment by jkodumal** [comment]
👍

File: demo/demo.py (hunk 3 above)
Comment (drichelson) [inline demo/demo.py]: I did a wholesale change- including removing the api_key prefix from the auth header.

File: ldclient/client.py [hunk 4]
```diff
@@ -222,26 +233,40 @@ def send_event(value):
...
-        if 'offVariation' in flag and flag['offVariation']:
-                value = _get_variation(flag, flag['offVariation'])
-                send_event(value)
-                return value
-
-        send_event(default)
+        send_event(default, flag.get('version'))
         return default
 
+    def all_flags(self, user):
```
Comment (drichelson) [inline ldclient/client.py]: @jkodumal take note of these checks when implementing in other sdks.

**Pull request merged**